ecosystem/nft-metadata-crawler/src/asset_uploader/throttler/mod.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    asset_uploader::worker::UploadRequest,
+    config::Server,
+    models::{
+        asset_uploader_request_statuses::AssetUploaderRequestStatuses,
+        asset_uploader_request_statuses_query::AssetUploaderRequestStatusesQuery,
+    },
+    schema,
+};
+use ahash::AHashSet;
+use anyhow::Context;
+use axum::{http::StatusCode as AxumStatusCode, response::IntoResponse, routing::post, Extension};
+use config::AssetUploaderThrottlerConfig;
+use diesel::{
+    r2d2::{ConnectionManager, Pool},
+    upsert::excluded,
+    ExpressionMethods, PgConnection, QueryDsl, RunQueryDsl,
+};
+use parking_lot::Mutex;
+use reqwest::{Client, StatusCode as ReqwestStatusCode};
+use serde::Deserialize;
+use std::{
+    collections::BTreeSet,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+use tokio::sync::Notify;
+use tracing::{debug, error};
+use url::Url;
+
+pub mod config;
+
+const FIVE_MINUTES: Duration = Duration::from_secs(60 * 5);
+
+// Structs below are for accessing relevant data in a typed way for Cloudflare API calls
+#[derive(Debug, Deserialize)]
+struct CloudflareImageUploadResponseResult {
+    id: String,
+}
+
+#[derive(Debug, Deserialize)]
+struct CloudflareImageUploadResponse {
+    errors: Vec<String>,
+    result: CloudflareImageUploadResponseResult,
+}
+
+#[derive(Clone)]
+pub struct AssetUploaderThrottlerContext {
+    config: AssetUploaderThrottlerConfig,
+    pool: Pool<ConnectionManager<PgConnection>>,
+    asset_queue: Arc<Mutex<BTreeSet<AssetUploaderRequestStatusesQuery>>>,
+    in_progress_assets: Arc<Mutex<AHashSet<AssetUploaderRequestStatusesQuery>>>,
```

Comment:
should we use single lock for both queues to avoid possible race conditions btw queue updates and uploads? where an asset could be added to the queue by `update_queue` just as it's being removed and processed by `handle_upload_assets`? chatgpt suggested https://docs.rs/crossbeam-queue/latest/crossbeam_queue/ as well

Reply:
oki doki